openhands-sdk/openhands/sdk/agent/base.py

```diff
@@ -196,13 +197,39 @@ def _initialize(self, state: "ConversationState"):
             return
 
         tools: list[ToolDefinition] = []
-        for tool_spec in self.tools:
-            tools.extend(resolve_tool(tool_spec, state))
 
-        # Add MCP tools if configured
-        if self.mcp_config:
-            mcp_tools = create_mcp_tools(self.mcp_config, timeout=30)
-            tools.extend(mcp_tools)
+        # Use ThreadPoolExecutor to parallelize tool resolution
+        with ThreadPoolExecutor(max_workers=4) as executor:
+            futures = []
+
+            # Submit tool resolution tasks
+            for tool_spec in self.tools:
+                future = executor.submit(resolve_tool, tool_spec, state)
+                futures.append(("tool", future))
+
+            # Submit MCP tools creation if configured
+            if self.mcp_config:
+                future = executor.submit(create_mcp_tools, self.mcp_config, 30)
+                futures.append(("mcp", future))
+
+            # Submit built-in tools creation
+            for tool_class in BUILT_IN_TOOLS:
+                future = executor.submit(tool_class.create, state)
+                futures.append(("builtin", future))
+
+            # Collect results as they complete
+            for future_type, future in futures:
+                try:
+                    result = future.result()
+                    if future_type == "tool":
+                        tools.extend(result)
+                    elif future_type == "mcp":
+                        tools.extend(result)
+                    elif future_type == "builtin":
+                        tools.extend(result)
```

Comment:
The case for different `future_type` is exactly the same, so it make sense to just remove the "future_type" here and simplify the code to be just:
1. add futures from tool and MCP
2. wait for future to finish, and add each one to the list when done
3. do tool filtering
4. add built-in tools with no thread pool